### PR TITLE
Add database init retry

### PR DESCRIPTION
**Pull request by gchudnov**
Hello guys!
bumped into #3114 issue.

Made a fix as described in the ticket.
Not sure about the right strategy for the # of retries and delay between them. Set to 3 retries with 5 seconds delay in-between.

If that should be resolved in a different way or needs any other changes, let me know.
Please review!

Thank you,
Grigoriy

### CODE REVIEW

**Comment by poppyseedplehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit bce34a25770d6cadae6f5aaff3e3f2b0feaf195c (Job results: [1380](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1380/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit bce34a25770d6cadae6f5aaff3e3f2b0feaf195c (Job results: [4500](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4500/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit bce34a25770d6cadae6f5aaff3e3f2b0feaf195c (Job results: [1403](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1403/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit bce34a25770d6cadae6f5aaff3e3f2b0feaf195c (Job results: [680](https://jenkins.osquery.io/job/osqueryPullRequestOSX/680/)) failed one or more tests (**macOS/OS X**).

File: osquery/core/init.cpp [hunk 1]
```diff
@@ -570,11 +573,25 @@ void Initializer::start() const {
...
     DatabasePlugin::setRequireWrite(tool_ == ToolType::DAEMON);
-    if (!DatabasePlugin::initPlugin()) {
-      LOG(ERROR) << RLOG(1629) << binary_
-                 << " initialize failed: Could not initialize database";
-      auto retcode = (isWorker()) ? EXIT_CATASTROPHIC : EXIT_FAILURE;
-      requestShutdown(retcode);
+
+    for (size_t i = 1; i <= kDatabaseMaxRetryCount; i++) {
+      auto ok = DatabasePlugin::initPlugin();
+      if (ok) {
```
Comment (jedi22) [inline osquery/core/init.cpp]: @theopolis Convention is ok.ok() right? How do you feel about this syntax?

Comment (theopolis) [inline osquery/core/init.cpp]: I'd recommend:
```cpp
if (DatabasePlugin::initPlugin().ok()) {
  break
}

File: osquery/core/init.cpp [hunk 2]
```diff
@@ -570,11 +573,25 @@ void Initializer::start() const {
...
+      }
+
+      if (i == kDatabaseMaxRetryCount) {
+        LOG(ERROR) << RLOG(1629) << binary_
+                   << " initialize failed: Could not initialize database";
+        auto retcode = (isWorker()) ? EXIT_CATASTROPHIC : EXIT_FAILURE;
+        requestShutdown(retcode);
+      }
+
+      LOG(WARNING) << RLOG(1629) << binary_
```
Comment (theopolis) [inline osquery/core/init.cpp]: I'd recommend removing these. If the goal is to wait a few moments for the database to be ready, you can instead check a lot faster, every 100/200ms for example, up until a max of 5 seconds have passed.

File: osquery/core/init.cpp (hunk 1 above)
Comment (gchudnov) [inline osquery/core/init.cpp]: got it, I'll fix it

File: osquery/core/init.cpp (hunk 2 above)
Comment (gchudnov) [inline osquery/core/init.cpp]: Yes, the idea was to wait a few moments for the database to be ready.
--
Sorry, not sure I get it... Would you like to remove `requestShutdown` and `LOG(WARNING)` ?
hm.. if `requestShutdown` is removed, is it OK to let the app running?

Comment (theopolis) [inline osquery/core/init.cpp]: Just the `LOG(WARNING)`, sorry for the ambiguity 👍

Comment (gchudnov) [inline osquery/core/init.cpp]: made some changes:
* changed inti status check to: `if (DatabasePlugin::initPlugin().ok()) { ..`
* reduced  the delay to 200 msec, increased the number of attempts to 25 (should give almost 5 seconds of waiting in total)
* removed warning and info logging

Let me know if it works.

Thank you!

**Comment by gchudnov** [comment]
Thank you!

**Pull request merged**